doc/developer/design/20230512_mir_cost_model.md

```diff
@@ -0,0 +1,149 @@
+- Storage (memory footprint/implicit storage)
+- Scalability (to what degree can this query be spread across workers)
+
+Our cost model should be *executable* and *accurate:*
+
+- Our cost model will be code that runs on actual MIR, producing a symbolic expression that represents the cost of the query along various dimensions. A user can use `EXPLAIN WITH(cost)`  to predict costs.
+- Our cost model should be tested on a variety of loads to ensure that our estimates aren’t total nonsense.
+
+# Motivation
+[motivation]: #motivation
+
+## Goals
+
+- Reify what we learn from various benchmarks/experiments.
+    - https://github.com/MaterializeInc/materialize/issues/17883
+    - https://github.com/MaterializeInc/materialize/issues/16511
+    - https://github.com/MaterializeInc/materialize/issues/15244
+    - https://github.com/MaterializeInc/materialize/issues/13140
+    - https://github.com/MaterializeInc/materialize/issues/12069
+    - https://github.com/MaterializeInc/materialize/issues/9157
+- Offer visibility/predicability.
+    - https://github.com/MaterializeInc/materialize/issues/14663
+    - https://github.com/MaterializeInc/materialize/issues/13046
+- Open opportunities for principled/well-informed choices in the optimizer.
+
+## Non-goals
+
+- Identify all opportunities to use the cost model in the optimizer.
+- Refactor the optimizer to be focused on the cost model.
+- Implement dynamic re-optimization based on changes in input statistics.
+- Collect all relevant statistics in storage and make them available to the cost model.
+
+# Explanation
+[explanation]: #explanation
+
+## Proposal
+
+A cost model is an abstract interpretation of the query: rather than running the query on relations and getting new relation out, we run the query on a symbolic description of its input relations and get a symbolic description of the result out.
+
+The general approach can be the same as `mz_transform::Typecheck`: recursively walk the `MirRelationExpr` with some contextual information, returning a symbolic term characterizing the query.
+
+### Context information
+
+Traditional databases track cardinality data and statistics about various columns. We will be interested in other statistics, too. Some possible examples:
+
+- How frequently does data arrive?
+- Is the given input physically monotonic or append-only? (I.e., there are no retractions.)
+- At what rate do we expect new data to arrive?
+- Some kind of locality property?
+
+### Symbolic costs
+
+A symbolic cost is a polynomial over the features of the input relations. For example, a batch query that loops over relation `R` twice and then computes its cross product with relation `Q` will have cost `2|R| + |R|*|Q|`. A traditional database typically computes cost just in terms of overall batch query time, but our costs are multi-dimensional. In rough feasibility order:
+
+| Dimension          | Input features                                               | Outputs                         |
```

Comment:
nit: As we move closer to the implementation it will be great if we can refine this table with the output type.